packages/flutter/lib/src/widgets/app.dart

```diff
@@ -689,35 +718,138 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
+      Locale userLocale = preferredLocales[localeIndex];
+      // Look for perfect match.
+      if (allSupportedLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode, userLocale.countryCode.hashCode))) {
+        return userLocale;
+      }
+      // Look for language+script match.
+      if (userLocale.scriptCode != null && languageAndScriptLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode))) {
+        return languageAndScriptLocales[ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode)];
+      }
+      // Look for language+country match.
+      if (userLocale.countryCode != null && languageAndCountryLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.countryCode.hashCode))) {
+        return languageAndCountryLocales[ui.hashValues(userLocale.languageCode.hashCode, userLocale.countryCode.hashCode)];
+      }
+      // If there was a languageCode-only match in the previous iteration's higher
+      // ranked preferred locale, we return it if the current userLocale does not
+      // have a perfect match.
+      if (matchesLanguageCode != null) {
+        return matchesLanguageCode;
+      }
+      // Look and store language-only match.
+      if (languageLocales.containsKey(userLocale.languageCode.hashCode)) {
+        matchesLanguageCode = languageLocales[userLocale.languageCode.hashCode];
+        // Since first (default) locale is usually highly preferred, we will allow
+        // a languageCode-only match to be instantly matched. If the next preferred
+        // languageCode is the same, we defer hastily returning until the next iteration
+        // since it can only be the same or improved match.
+        if (localeIndex == 0
+          && !(localeIndex + 1 < preferredLocales.length
+              && preferredLocales[localeIndex + 1].languageCode == userLocale.languageCode)) {
+          return matchesLanguageCode;
+        }
+      }
+      // countryCode only match. When all else except default supported locale fails,
+      // attempt to match by country only, as a user is likely to be familar with a
+      // language from their listed country.
+      if (userLocale.countryCode != null && countryLocales.containsKey(userLocale.countryCode.hashCode)) {
+        matchesCountryCode ??= countryLocales[userLocale.countryCode.hashCode];
+      }
+    }
+    // When there is no languageCode only match. Fallback to matching countryCode only. Country
+    // fallback only applies on iOS. When there is no countryCode only match, we return first
+    // suported locale.
+    Locale resolvedLocale = matchesLanguageCode ?? matchesCountryCode ?? supportedLocales.first;
+    if (matchesLanguageCode == null) {
+      debugPrint('Warning: Resolved locale (' + resolvedLocale.toString() + ') has a different language code than the user\'s preferred locales (' + preferredLocales.toString() + ').');
```

Comment:
@HansMuller Is a message like this what you had in mind?

Reply:
we shouldn't print warnings. We should either fail, or we should silently succeed. Warnings will just result in spam to the console that users then ignore and then they'll miss important messages.

Reply:
I'll take this out of this patch for now and if @HansMuller would like to land something of this nature later, we can do it separately.